File: Samples/Evaluate/how-to-get-data-validation-errors-count-t559295-17.1.3-/VB/DXApplication15/CustomCommandFactoryService.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraSpreadsheet
Imports DevExpress.XtraSpreadsheet.Commands
Imports DevExpress.XtraSpreadsheet.Services
Imports System
Imports System.Linq
Imports System.Windows.Forms

Namespace DXApplication15
	Public Class CustomCommandFactoryService
		Inherits SpreadsheetCommandFactoryServiceWrapper
		Public Sub New(ByVal service As ISpreadsheetCommandFactoryService)
			MyBase.New(service)
		End Sub
		Private privateControl As SpreadsheetControl
		Public Property Control() As SpreadsheetControl
			Get
				Return privateControl
			End Get
			Set(ByVal value As SpreadsheetControl)
				privateControl = value
			End Set
		End Property

		Public Overrides Function CreateCommand(ByVal id As SpreadsheetCommandId) As SpreadsheetCommand
            If id = SpreadsheetCommandId.DataToolsCircleInvalidData Then
                Return New CustomCircleInvalidDataCommand(Control)
            End If

            Return MyBase.CreateCommand(id)
		End Function

	End Class
	Public Class CustomCircleInvalidDataCommand
		Inherits CircleInvalidDataCommand
		Public Sub New(ByVal control As ISpreadsheetControl)
			MyBase.New(control)
		End Sub
		Private count As Integer = 0
		Public ReadOnly Property ErrorsCount() As Integer
			Get
				Return count
			End Get
		End Property
		Protected Overrides Sub ExecuteCore()
			MyBase.ExecuteCore()
			count = Me.ActiveSheet.InvalidDataCircles.Count
			MessageBox.Show("Data valiadtion errors count = " & count)
		End Sub
	End Class
End Namespace

File: Samples/Evaluate/how-to-get-data-validation-errors-count-t559295-17.1.3-/CS/DXApplication15/CustomCommandFactoryService.cs
using DevExpress.XtraSpreadsheet;
using DevExpress.XtraSpreadsheet.Commands;
using DevExpress.XtraSpreadsheet.Services;
using System;
using System.Linq;
using System.Windows.Forms;

namespace DXApplication15
{
    public class CustomCommandFactoryService : SpreadsheetCommandFactoryServiceWrapper
    {
        public CustomCommandFactoryService(ISpreadsheetCommandFactoryService service)
            : base(service)
        {
        }
        public SpreadsheetControl Control
        {
            get;
            set;
        }

        public override SpreadsheetCommand CreateCommand(SpreadsheetCommandId id)
        {
            if (id == SpreadsheetCommandId.DataToolsCircleInvalidData)
                return new CustomCircleInvalidDataCommand(Control);

            return base.CreateCommand(id);
        }

    }
    public class CustomCircleInvalidDataCommand : CircleInvalidDataCommand
    {
        public CustomCircleInvalidDataCommand(ISpreadsheetControl control)
            : base(control)
        {
        }
        int count = 0;
        public int ErrorsCount
        {
            get
            {
                return count;
            }
        }
        protected override void ExecuteCore()
        {
            base.ExecuteCore();
            count = this.ActiveSheet.InvalidDataCircles.Count;
            MessageBox.Show("Data valiadtion errors count = " + count);
        }
    }
}